File: secure-code-review/vulnerable-examples/python/secrets/hardcoded_secrets.py
"""
Hardcoded Secrets — vulnerable and secure patterns, side by side.

Scenario: a notification service that posts messages to a third-party
vendor and reads customer records from a database. It needs two
credentials: the vendor API key and a database password.

The vulnerable version is the pattern secret-scanners catch ~every
week in real repos. The secure version demonstrates the progression
most teams go through: env vars for local dev, a real secrets manager
in production, with a single `get_secret()` abstraction so code does
not care which environment it is running in.

Important: if you find hardcoded credentials during a review, they are
compromised. git history preserves them forever. Removing the line is
not a fix — rotate the secret at the source of truth.
"""

import os
from functools import lru_cache
from typing import Optional

import requests


# =============================================================================
# VULNERABLE — do not do any of this.
# =============================================================================

# Problem 1: credentials in source. Visible to anyone with read access
# to the repo (including every employee, every CI runner, every LLM
# that indexed it). Lives in git history forever — `git rm` does not
# remove it from the log.
VENDOR_API_KEY = "sk_live_EXAMPLE_NOT_A_REAL_KEY

# Problem 2: password embedded in a connection string, committed.
DATABASE_URL = "postgresql://svc_notify:S3cretP@ssw0rd@prod-db.internal:5432/notifications"

# Problem 3: a "real-looking fallback" in env-var defaults. Looks like
# defensive coding; actually a hardcoded credential that kicks in if
# the env var is missing.
SLACK_WEBHOOK = os.environ.get(
    "SLACK_WEBHOOK",
    "https://hooks.slack.com/services/EXAMPLE/WEBHOOK/URL
)


def send_notification_vulnerable(message: str) -> None:
    requests.post(
        "https://api.notify.example.com/v1/send",
        json={"msg": message},
        headers={"Authorization": f"Bearer {VENDOR_API_KEY}"},
        timeout=5,
    )


# =============================================================================
# SECURE — single abstraction, fail-fast on startup, prod uses a KMS-
# backed secrets manager, dev uses environment variables.
# =============================================================================


class SecretsProvider:
    """
    One interface over local env vars and cloud secrets managers. The
    abstraction lets application code call `secrets.get("vendor_api_key")`
    without caring whether we are in dev, staging, or prod.

    Two properties of this design that matter for review:

      - `lru_cache` on the fetch — secrets managers charge per call
        and have rate limits. Cache per process.

      - No default values. If the secret is missing we raise, loud and
        early, at application startup. A silent fallback to "" or to a
        placeholder string hides configuration bugs until they become
        incidents.
    """

    def __init__(self, environment: Optional[str] = None):
        self.environment = environment or os.environ.get("ENV", "development")
        self._client = None
        if self.environment == "production":
            import boto3  # imported lazily so dev does not need boto3
            self._client = boto3.client("secretsmanager")

    @lru_cache(maxsize=32)
    def get(self, name: str) -> str:
        if self.environment == "production":
            return self._get_from_aws(name)
        return self._get_from_env(name)

    def _get_from_aws(self, name: str) -> str:
        # Secret IDs follow a convention — `/app/<env>/<name>`.
        secret_id = f"/notify/prod/{name}"
        resp = self._client.get_secret_value(SecretId=secret_id)
        value = resp.get("SecretString")
        if not value:
            raise RuntimeError(f"secret {secret_id} is empty or binary")
        return value

    def _get_from_env(self, name: str) -> str:
        # Env var names are upper-snake-case versions of the secret name.
        env_name = name.upper()
        value = os.environ.get(env_name)
        if not value:
            raise RuntimeError(
                f"secret {name} is missing. Set {env_name} in your .env file "
                f"(see .env.example for the required keys)."
            )
        return value


# Instantiate once at module load. Application code takes a dependency
# on the abstraction, not on the underlying provider.
secrets = SecretsProvider()


def send_notification_secure(message: str) -> None:
    # No credential is ever written to source. The key is fetched from
    # Secrets Manager in prod, from the local env in dev, and a missing
    # value fails fast rather than using a placeholder.
    api_key = secrets.get("vendor_api_key")
    requests.post(
        "https://api.notify.example.com/v1/send",
        json={"msg": message},
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=5,
    )


# =============================================================================
# If you inherit a repo with committed secrets, this is the response:
#
#   1. Treat the secret as compromised. Do not "just remove the line."
#   2. Rotate the credential at the source of truth (vendor, database,
#      cloud IAM). The new value goes into Secrets Manager, not the
#      repo.
#   3. Audit usage of the old credential before rotation. You want to
#      know whether it was used from an unexpected source in the log
#      window — that answers "did we get breached via this leak."
#   4. Purge the secret from git history only after rotation. Tools:
#      BFG Repo-Cleaner, git-filter-repo. Force-push with team
#      coordination.
#   5. Add a pre-commit hook (gitleaks, detect-secrets) so the next
#      developer's mistake is caught before it lands.
#
# A checklist without step 1 is a liability — it gives the team a
# false sense of closure while the credential is still live.
# =============================================================================

File: secure-code-review/vulnerable-examples/javascript/secrets/hardcoded_secrets.js
/**
 * Hardcoded Secrets — vulnerable and secure patterns, side by side.
 *
 * Scenario: a Node.js service that charges cards via Stripe and
 * writes transaction records to Postgres. It needs two credentials:
 * the Stripe secret key and the Postgres connection string.
 *
 * The vulnerable version is the pattern `gitleaks` flags every day.
 * The secure version is the progression: dotenv-safe for local dev,
 * a secrets manager abstraction for production, one config module
 * that validates the loaded values and fails fast on startup.
 *
 * Reminder: if you find committed secrets during review, treat them
 * as compromised. Removing the line does not un-leak them — they
 * are in git history, on every developer's machine, in every CI
 * cache, and very possibly in training data for some LLM.
 */

// ===========================================================================
// VULNERABLE — do not do any of this.
// ===========================================================================

// Bug 1: a literal secret in source. gitleaks catches this on commit,
// but only if the hook is installed. Lives in git history either way.
const STRIPE_SECRET = "sk_live_EXAMPLE_NOT_A_REAL_KEY // example format

// Bug 2: credentials embedded in a connection URL, committed.
const DATABASE_URL =
  "postgres://svc_payments:S3cretP@ssw0rd@prod-db.internal:5432/payments";

// Bug 3: "defensive" fallback default that is itself a live secret.
// Reads plausible — ships a real credential if the env var is missing.
const DATADOG_API_KEY =
  process.env.DATADOG_API_KEY || "a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6";

// Bug 4: test fixture with real-shaped credentials that grep-bait the
// scanner into false positives, so the real leak hides in the noise.
// (Yes, I have seen this used intentionally. It is a bad pattern.)
const TEST_STRIPE_KEY = "sk_test_FAKE0000000000000000000000";

async function chargeVulnerable(amountCents, sourceToken) {
  const Stripe = require("stripe");
  const stripe = new Stripe(STRIPE_SECRET);
  return stripe.charges.create({ amount: amountCents, source: sourceToken, currency: "usd" });
}


// ===========================================================================
// SECURE — one config module, fail-fast validation, pluggable source.
// ===========================================================================
//
// Directory convention:
//
//   .env.example      — committed. Lists the variable NAMES with
//                       placeholder values. dotenv-safe uses this as
//                       the "required keys" contract.
//
//   .env              — NOT committed (in .gitignore). Real values
//                       for local development only.
//
//   Production        — env vars come from the platform's secret
//                       store (AWS Secrets Manager via a
//                       pre-deploy hook, HashiCorp Vault agent,
//                       Kubernetes externalSecrets, etc.). The
//                       application code is unchanged — it reads
//                       from process.env either way.
//
// ---------------------------------------------------------------------------
// .env.example (committed, no real values) — lists required vars:
//
//   NODE_ENV=
//   STRIPE_SECRET=
//   DATABASE_URL=
//   DATADOG_API_KEY=
//
// ---------------------------------------------------------------------------

const { z } = require("zod");

// In CI / prod, `.env` does not exist — dotenv-safe is a no-op
// because real env vars are already populated. Locally it ensures
// every required key is present and yells if one is missing.
require("dotenv-safe").config({
  allowEmptyValues: false,
  example: ".env.example",
});

// Schema-driven validation: every environment variable is declared
// with the type + constraints it must satisfy. If something is
// missing or malformed, the process crashes on startup with a clear
// error — not three hours later when a customer hits a broken code
// path.
const EnvSchema = z.object({
  NODE_ENV: z.enum(["development", "test", "production"]),

  // Format hint catches swapped keys (live vs test) and typos.
  STRIPE_SECRET: z
    .string()
    .regex(/^sk_(live|test)_[A-Za-z0-9]{16,}$/, "invalid Stripe secret format"),

  DATABASE_URL: z.string().url(),

  // Hex, exactly 32 chars — catches truncated or malformed keys.
  DATADOG_API_KEY: z.string().regex(/^[0-9a-f]{32}$/),
});

let env;
try {
  env = EnvSchema.parse(process.env);
} catch (err) {
  // Fail fast with the clearest possible diagnostic.
  console.error("Environment validation failed:");
  for (const issue of err.errors ?? []) {
    console.error(`  - ${issue.path.join(".")}: ${issue.message}`);
  }
  process.exit(1);
}

// Guardrail: in production, refuse to run if a Stripe test key is
// loaded, or vice versa. Catches deploy-time config mixups.
if (env.NODE_ENV === "production" && env.STRIPE_SECRET.startsWith("sk_test_")) {
  console.error("Refusing to start: Stripe TEST key loaded in production");
  process.exit(1);
}

module.exports.env = env;


// ===========================================================================
// Secrets-manager abstraction — optional, and preferred over raw env
// vars for high-value secrets that should be rotatable without a
// redeploy.
// ===========================================================================

class SecretsProvider {
  constructor(environment) {
    this.environment = environment;
    this.cache = new Map();
    if (environment === "production") {
      const {
        SecretsManagerClient,
        GetSecretValueCommand,
      } = require("@aws-sdk/client-secrets-manager");
      this.client = new SecretsManagerClient({});
      this.GetSecretValueCommand = GetSecretValueCommand;
    }
  }

  async get(name) {
    if (this.cache.has(name)) return this.cache.get(name);

    let value;
    if (this.environment === "production") {
      const resp = await this.client.send(
        new this.GetSecretValueCommand({ SecretId: `/payments/prod/${name}` })
      );
      value = resp.SecretString;
      if (!value) {
        throw new Error(`secret ${name} is empty or binary`);
      }
    } else {
      const envName = name.toUpperCase();
      value = process.env[envName];
      if (!value) {
        throw new Error(
          `secret ${name} missing; set ${envName} in your .env file`
        );
      }
    }
    this.cache.set(name, value);
    return value;
  }
}

const secrets = new SecretsProvider(env.NODE_ENV);

async function chargeSecure(amountCents, sourceToken) {
  const Stripe = require("stripe");

  // Lazy fetch — prod hits Secrets Manager once, then cache.
  const stripe = new Stripe(await secrets.get("stripe_secret"));
  return stripe.charges.create({
    amount: amountCents,
    source: sourceToken,
    currency: "usd",
  });
}


// ===========================================================================
// If you inherit a repo with committed secrets, this is the response:
//
//   1. Treat every committed secret as compromised. Removing the line
//      is not remediation.
//
//   2. Rotate the credential at its source (Stripe dashboard, AWS
//      IAM, DB user) and put the new value in the secrets manager.
//
//   3. Audit usage in the log window before rotation. Unknown callers
//      or anomalous patterns = incident.
//
//   4. Purge from history after rotation, with git-filter-repo or BFG,
//      coordinating a force-push with the whole team.
//
//   5. Install gitleaks (or equivalent) as a pre-commit hook and as a
//      CI step on the default branch. The next developer's mistake is
//      then caught before the credential is ever pushed.
// ===========================================================================

module.exports.chargeSecure = chargeSecure;
module.exports.secrets = secrets;
